Python/flowgraph.c

```diff
@@ -807,35 +824,30 @@ calculate_stackdepth(cfg_builder *g)
         basicblock *next = b->b_next;
         for (int i = 0; i < b->b_iused; i++) {
             cfg_instr *instr = &b->b_instr[i];
-            int effect = stack_effect(instr->i_opcode, instr->i_oparg, 0);
-            if (effect == PY_INVALID_STACK_EFFECT) {
+            stack_effects effects;
+            if (get_stack_effects(instr->i_opcode, instr->i_oparg, 0, &effects) < 0) {
                 PyErr_Format(PyExc_SystemError,
                              "Invalid stack effect for opcode=%d, arg=%i",
                              instr->i_opcode, instr->i_oparg);
                 goto error;
             }
-            int new_depth = depth + effect;
+            int new_depth = depth + effects.net;
             if (new_depth < 0) {
-               PyErr_Format(PyExc_ValueError,
-                            "Invalid CFG, stack underflow");
-               goto error;
-            }
-            if (new_depth > maxdepth) {
-                maxdepth = new_depth;
+                PyErr_Format(PyExc_ValueError,
+                             "Invalid CFG, stack underflow");
+                goto error;
             }
+            maxdepth = Py_MAX(maxdepth, depth + effects.max);
```

Comment:
It looks like we used to allocate `MAX(depth + effects.net)` for the stack and now we're doing `MAX(depth + effects.max)`. Were we under allocating before?

Reply:
No, the maximum stack depth was calculated correctly based on the stack effect of the instruction. However, this meant that it wasn't always possible to pass values on the stack between uops. Consider this contrived example:

```
op(A, (-- val1, val2)) {}
op(B, (val1, val2 -- )) {}

macro(OP1) = A + B;

inst(OP, (--)) {}
        
family(OP, 0) = { OP1 };
```

We have an instruction, `OP`, that we'd like to specialize with `OP1`. Both forms have the same stack effect, `0`, but the implementation of `OP1` wants to be able to pass two values temporarily on the stack (from `A` to `B`). This kind of construction was impossible with the previous way we calculated the maximum stack depth.